cms/djangoapps/contentstore/tasks.py

```diff
@@ -1066,3 +1070,186 @@ def undo_all_library_source_blocks_ids_for_course(course_key_string, v1_to_v2_li
             store.update_item(draft_library_source_block, None)
     # return success
     return
+
+
+class CourseLinkCheckTask(UserTask):  # pylint: disable=abstract-method
+    """
+    Base class for course link check tasks.
+    """
+
+    @staticmethod
+    def calculate_total_steps(arguments_dict):
+        """
+        Get the number of in-progress steps in the link check process, as shown in the UI.
+
+        For reference, these are:
+        1. Scanning
+        2. Verifying
+        """
+        return 2
+
+    @classmethod
+    def generate_name(cls, arguments_dict):
+        """
+        Create a name for this particular task instance.
+
+        Arguments:
+            arguments_dict (dict): The arguments given to the task function
+
+        Returns:
+            str: The generated name
+        """
+        key = arguments_dict['course_key_string']
+        return f'Broken link check of {key}'
+
+
+@shared_task(base=CourseLinkCheckTask, bind=True)
+def check_broken_links(self, user_id, course_key_string, language):
+    """
+    Checks for broken links in a course. Store the results in a file.
+    """
+    courselike_key = CourseKey.from_string(course_key_string)
+
+    try:
+        user = User.objects.get(pk=user_id)
+    except User.DoesNotExist:
+        with translation_language(language):
+            self.status.fail(UserErrors.UNKNOWN_USER_ID.format(user_id))
+        return
+    if not has_course_author_access(user, courselike_key):
+        with translation_language(language):
+            self.status.fail(UserErrors.PERMISSION_DENIED)
+        return
+    
+    courselike_block = modulestore().get_course(courselike_key)
```

Comment:
This method's code up to and including this line is all about gaining access to course content. By contrast, what follows is all about searching for broken links within that content. From a single responsibility viewpoint, this is a sign that you want to modularize this work into two distinct methods. Note that from a unit testing perspective, single responsibility is your friend, as it's much clearer to devise separate unit tests for what's above and what's below than it is to come up with a unit test for the function as is. Even if you're not writing the unit tests now, it's a good idea to structure your code to make it easy to write unit tests sometime.